### DISCUSSION TITLE
async evaluateLater inside effect can cause infinite loop

### DISCUSSION THREAD

**Discussion by nikitaourazbaev** [alpinejs/alpine#3558] [1. Help] [answered]
I'm running into an odd behaviour with `await`ing functions, `evaluateLater`, and Alpine’s `effect`. This happens inside a custom directive but I’ve also tested it outside Alpine (using `Alpine.effect` and accessing the element’s data stack with `$data` or with `mergeProxies(closestDataStack(el))`)

Here’s a reduced test case:

https://codepen.io/magiclantern/pen/KKGZOab/619b184631abeed14bbff119f8839c2e

In my real code, the `getContent` function is a `fetch` call to a server and `runBefore` and `runAfter` are Alpine expressions from elsewhere. But otherwise this is an accurate representation of the issue I’m running into. Once the commented-out line in `runBefore` is un-commented, the code will go into an infinite loop. 

The key to getting this to happen seems to be this:

If the "before" and "after" function are setting / getting a property in the same data stack as the property causing the `effect`, it can cause that property to become reevaluated, running the `effect` again, and causing an infinite loop. 

For some reason this doesn’t happen with `this.$root`, but it does if accessing `this.processing` in the "before" function.

I’ve tried running some or all of the functions inside `dontAutoEvaluateFunctions` and also inside `disableEffectScheduling` but as long as 

1. one of them is `await`ed, and
2. both the "before" and "after" methods access that `processing` property in the data stack 

It will trigger the original `effect` over and over.

I’ve even tried having `effect` do nothing but dispatch an event, listen for the event outside of Alpine, and call those methods from outside, and this still happens.

I feel like I’m probably misunderstanding something about `effect` but I’m not sure how to debug this further. Why does accessing `this.processing` cause the initial property to act as though it's changed?

**Comment by ekwoka** [comment, marked as answer]
The best way to solve this is to rearchitect to not modify any of the effects dependencies in the effect.

If that can't be handled, just do what `$watch` does and check the dependencies that you do care about, and then tossing the operation into a `queueMicrotask`


It's a bit hard to say what would really be best for your situation, since it would take some refactoring and your example may differ meaningfully from what the goal is.

I'd probably adjust it to specifically be watching the dependent variable, and not the whole function.

**Reply by nikitaourazbaev** [reply]
@ekwoka thank you, that was very helpful.

> The best way to solve this is to rearchitect to not modify any of the effects dependencies in the effect.

I’m trying to do that, I tried having the effect only set the value and dispatch an event which was handled outside Alpine as my assumption was `effect` reacts to anything touched inside the immediate scope of the function passed into it. But maybe the timing within the event loop also matters and even if the properties are touched outside that function they can still re-trigger the effect? I’ve tried looking into this deeper in both Alpine’s `reactivity.js` and `@vue/reactivity` but haven’t yet found a definitive answer. I may need to experiment more with this.

> do what $watch does and check the dependencies that you do care about, and then tossing the operation into a queueMicrotask

Yes, `queueMicrotask` does seem to work here! I will use that for now while I explore how best to rearchitect. Thanks again for your help.

**Reply by SimoTod** [reply]
I think it's because you got confused by evaluateLater. 
The function just converts a string to the real javascript code but it doesn't access variables in advance and it's not linked to the reactivity. 
In your example, you are still running the function inside the effect therefore every variable you touch is tracked.

**Reply by nikitaourazbaev** [reply]
@SimoTod Thank you for your reply! 

> In your example, you are still running the function inside the effect therefore every variable you touch is tracked.

I think I understand that now, but I’m still not 100% clear on what 'inside the effect' means exactly.

In the example I posted above the `evaluateLater`s happen inside the `effect` but if I move them into a callback function:

https://codepen.io/magiclantern/pen/KKGQVgG/79111484521ac4b406363fc3934c9001

Would this still be considered 'inside the effect'? If I don't run the callback inside `queueMicrotask` the infinite loop will happen here too. (I based this on what `$watch` does as per @ekwoka’s suggestion.)

If I only dispatch an event inside the effect:

https://codepen.io/magiclantern/pen/yLRvrjw/2479217d1a5ac78a1979aa808262bd89

And run the callback in an event handler, the infinite loop would still happen if I don’t dispatch the callback inside a `queueMicrotask`.

Can `effect`’s tracking reach into that event handler?

Because `queueMicrotask` avoids the infinite loop in both cases, it feels like there’s a timing element at play. If I console log the expression’s journey from `evaluateLater` through Alpine’s evaluator at some point it gets converted to an async function, and I wonder if that might be at play?

**Reply by ekwoka** [reply]
Basically, anything that runs synchronously in the effect is tracked.

**Reply by SimoTod** [reply]
No, the fact that the function is async doesn't matter.

Alpine.effect is simply doing:
1 start looking for variables to be tracked
2 execute the js code
3 stop looking for variables to be tracked

You can add as many indirections as you like using callbacks but as long as your code runs between point 1 and 3, variables will still be tracked. 

queueMicrotask (to make it simpler) means that the code is not executed at that point but later when Alpine.effect has already stopped looking for variables to track. You would get the same result with a setTimeout. 

So evaluateLater just creates a callable function from a string and you call that function

**Reply by ekwoka** [reply]
To clarify, Async functions still execute synchronously until the first await. The resolution of the async function (and then anything awaiting it or any `.then` callbacks) happens in the microTask queue.

**Reply by nikitaourazbaev** [reply]
Thanks so much to you both for taking the time to explain this, it has really helped clear things up.
